File: medihub-frontend/src/app/services/rendezvous.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { RendezVous, DisponibiliteMedecin, HistoriqueRendezVous } from '../models/models';

@Injectable({ providedIn: 'root' })
export class RendezVousService {
  private api = '/api/rendezvous';

  constructor(private http: HttpClient) {}

  getAll(): Observable<RendezVous[]> { return this.http.get<RendezVous[]>(this.api); }
  getById(id: number): Observable<RendezVous> { return this.http.get<RendezVous>(`${this.api}/${id}`); }
  getByPatient(patientId: number): Observable<RendezVous[]> { return this.http.get<RendezVous[]>(`${this.api}/patient/${patientId}`); }
  getByMedecin(medecinId: number): Observable<RendezVous[]> { return this.http.get<RendezVous[]>(`${this.api}/medecin/${medecinId}`); }
  getProchainsByMedecin(medecinId: number): Observable<RendezVous[]> { return this.http.get<RendezVous[]>(`${this.api}/medecin/${medecinId}/prochains`); }
  getByDate(date: string): Observable<RendezVous[]> { return this.http.get<RendezVous[]>(`${this.api}/date/${date}`); }
  create(r: RendezVous): Observable<RendezVous> { return this.http.post<RendezVous>(this.api, r); }
  update(id: number, r: RendezVous): Observable<RendezVous> { return this.http.put<RendezVous>(`${this.api}/${id}`, r); }
  delete(id: number): Observable<void> { return this.http.delete<void>(`${this.api}/${id}`); }
  changerStatut(id: number, statut: string, acteur: string): Observable<RendezVous> {
    return this.http.patch<RendezVous>(`${this.api}/${id}/statut`, null, {
      params: new HttpParams().set('statut', statut).set('acteur', acteur)
    });
  }
  getHistorique(id: number): Observable<HistoriqueRendezVous[]> { return this.http.get<HistoriqueRendezVous[]>(`${this.api}/${id}/historique`); }

  // Disponibilités
  getDisponibilitesMedecin(medecinId: number): Observable<DisponibiliteMedecin[]> { return this.http.get<DisponibiliteMedecin[]>(`${this.api}/disponibilites/medecin/${medecinId}`); }
  getDisponibilitesDate(date: string): Observable<DisponibiliteMedecin[]> { return this.http.get<DisponibiliteMedecin[]>(`${this.api}/disponibilites/date/${date}`); }
  addDisponibilite(d: DisponibiliteMedecin): Observable<DisponibiliteMedecin> { return this.http.post<DisponibiliteMedecin>(`${this.api}/disponibilites`, d); }
  deleteDisponibilite(id: number): Observable<void> { return this.http.delete<void>(`${this.api}/disponibilites/${id}`); }
}

File: rendezvous-service/src/main/java/com/examplemicro/rendezvousservice/Services/RendezVousService.java
package com.examplemicro.rendezvousservice.Services;

import com.examplemicro.rendezvousservice.Entities.DisponibiliteMedecin;
import com.examplemicro.rendezvousservice.Entities.HistoriqueRendezVous;
import com.examplemicro.rendezvousservice.Entities.RendezVous;
import com.examplemicro.rendezvousservice.Exception.RendezVousNotFoundException;
import com.examplemicro.rendezvousservice.Repositories.DisponibiliteMedecinRepository;
import com.examplemicro.rendezvousservice.Repositories.HistoriqueRendezVousRepository;
import com.examplemicro.rendezvousservice.Repositories.RendezVousRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import com.examplemicro.rendezvousservice.dto.RendezVousEnrichiDTO;
import com.examplemicro.rendezvousservice.dto.UtilisateurDTO;
import com.examplemicro.rendezvousservice.client.UtilisateurClient;
@Service
@Transactional
public class RendezVousService {

    private final RendezVousRepository rendezVousRepository;
    private final DisponibiliteMedecinRepository disponibiliteRepository;
    private final HistoriqueRendezVousRepository historiqueRepository;
    private final UtilisateurClient utilisateurClient;
    private final NotificationProducer notificationProducer;

    public RendezVousService(
            RendezVousRepository rendezVousRepository,
            DisponibiliteMedecinRepository disponibiliteRepository,
            HistoriqueRendezVousRepository historiqueRepository,
            UtilisateurClient utilisateurClient,
            NotificationProducer notificationProducer) {    // ← ADD THIS
        this.rendezVousRepository = rendezVousRepository;
        this.disponibiliteRepository = disponibiliteRepository;
        this.historiqueRepository = historiqueRepository;
        this.utilisateurClient = utilisateurClient;
        this.notificationProducer = notificationProducer;   // ← ADD THIS
    }
    // ─── RENDEZ-VOUS ────────────────────────────────────────────────────────────

    public List<RendezVous> getAllRendezVous() {
        return rendezVousRepository.findAll();
    }

    public RendezVous getRendezVousById(Long id) {
        return rendezVousRepository.findById(id)
                .orElseThrow(() -> new RendezVousNotFoundException(id));
    }

    public List<RendezVous> getRendezVousByPatient(Long patientId) {
        return rendezVousRepository.findByPatientId(patientId);
    }

    public List<RendezVous> getRendezVousByMedecin(Long medecinId) {
        return rendezVousRepository.findByMedecinId(medecinId);
    }

    public List<RendezVous> getRendezVousByDate(LocalDate date) {
        return rendezVousRepository.findByDateRendezVous(date);
    }

    public List<RendezVous> getRendezVousByStatut(RendezVous.StatutRendezVous statut) {
        return rendezVousRepository.findByStatut(statut);
    }

    public List<RendezVous> getProchainsByMedecin(Long medecinId) {
        return rendezVousRepository.findProchainsByMedecin(medecinId, LocalDate.now());
    }

    public RendezVous createRendezVous(RendezVous rendezVous) {
        RendezVous saved = rendezVousRepository.save(rendezVous);
        enregistrerHistorique(saved.getId(), "CREATION", null,
                saved.getStatut().name(), "Rendez-vous créé", saved.getCreateur());
        notificationProducer.envoyerNotification(saved, "RDV_CREE");
        return saved;
    }

    public RendezVous updateRendezVous(Long id, RendezVous updated) {
        RendezVous existing = getRendezVousById(id);
        String ancienStatut = existing.getStatut().name();

        existing.setPatientId(updated.getPatientId());
        existing.setMedecinId(updated.getMedecinId());
        existing.setDateRendezVous(updated.getDateRendezVous());
        existing.setHeureRendezVous(updated.getHeureRendezVous());
        existing.setMotif(updated.getMotif());
        existing.setNotes(updated.getNotes());
        existing.setStatut(updated.getStatut());

        RendezVous saved = rendezVousRepository.save(existing);
        enregistrerHistorique(id, "MODIFICATION", ancienStatut,
                saved.getStatut().name(), "Rendez-vous modifié", updated.getCreateur());
         return saved;
    }

    public RendezVous changerStatut(Long id, RendezVous.StatutRendezVous nouveauStatut, String acteur) {
        RendezVous existing = getRendezVousById(id);
        String ancienStatut = existing.getStatut().name();
        existing.setStatut(nouveauStatut);
        RendezVous saved = rendezVousRepository.save(existing);

        // historique
        enregistrerHistorique(id, "CHANGEMENT_STATUT", ancienStatut,
                nouveauStatut.name(), "Statut modifié par " + acteur, acteur);

        // notification RabbitMQ
        notificationProducer.envoyerNotification(saved, "RDV_" + nouveauStatut.name());

        return saved;
    }

    public void deleteRendezVous(Long id) {
        RendezVous existing = getRendezVousById(id);
        enregistrerHistorique(id, "SUPPRESSION", existing.getStatut().name(),
                null, "Rendez-vous supprimé", "SYSTEM");
        rendezVousRepository.delete(existing);
    }

    // ─── DISPONIBILITES ─────────────────────────────────────────────────────────

    public List<DisponibiliteMedecin> getDisponibilitesByMedecin(Long medecinId) {
        return disponibiliteRepository.findByMedecinId(medecinId);
    }

    public List<DisponibiliteMedecin> getDisponibilitesByMedecinAndDate(Long medecinId, LocalDate date) {
        return disponibiliteRepository.findByMedecinIdAndDateDisponibilite(medecinId, date);
    }

    public List<DisponibiliteMedecin> getDisponiblesParDate(LocalDate date) {
        return disponibiliteRepository.findByDateDisponibiliteAndDisponibleTrue(date);
    }

    public DisponibiliteMedecin addDisponibilite(DisponibiliteMedecin disponibilite) {
        return disponibiliteRepository.save(disponibilite);
    }

    public DisponibiliteMedecin updateDisponibilite(Long id, DisponibiliteMedecin updated) {
        DisponibiliteMedecin existing = disponibiliteRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Disponibilité introuvable: " + id));
        existing.setDateDisponibilite(updated.getDateDisponibilite());
        existing.setHeureDebut(updated.getHeureDebut());
        existing.setHeureFin(updated.getHeureFin());
        existing.setDisponible(updated.isDisponible());
        return disponibiliteRepository.save(existing);
    }

    public void deleteDisponibilite(Long id) {
        disponibiliteRepository.deleteById(id);
    }

    // ─── HISTORIQUE ─────────────────────────────────────────────────────────────

    public List<HistoriqueRendezVous> getHistoriqueByRendezVous(Long rendezVousId) {
        return historiqueRepository.findByRendezVousIdOrderByDateActionDesc(rendezVousId);
    }

    private void enregistrerHistorique(Long rendezVousId, String action,
                                       String ancienStatut, String nouveauStatut,
                                       String details, String effectuePar) {
        HistoriqueRendezVous historique = new HistoriqueRendezVous();
        historique.setRendezVousId(rendezVousId);
        historique.setAction(action);
        historique.setAncienStatut(ancienStatut);
        historique.setNouveauStatut(nouveauStatut);
        historique.setDetails(details);
        historique.setEffectuePar(effectuePar);
        historiqueRepository.save(historique);
    }
    public RendezVousEnrichiDTO getRendezVousEnrichi(Long id) {
        RendezVous rdv = getRendezVousById(id);
        UtilisateurDTO patient = utilisateurClient.getUserById(rdv.getPatientId());
        UtilisateurDTO medecin = utilisateurClient.getUserById(rdv.getMedecinId());
        return new RendezVousEnrichiDTO(rdv, patient, medecin);
    }
}
